include/range/v3/span.hpp

```diff
@@ -0,0 +1,256 @@
+            using index_type = std::ptrdiff_t;
+            using pointer = T*;
+            using reference = T&;
+            using iterator = T*;
+            using reverse_iterator = ranges::reverse_iterator<iterator>;
+
+            static constexpr index_type extent = N;
+
+            // [span.cons], span constructors, copy, assignment, and destructor
+            constexpr span() = default;
+            constexpr span(std::nullptr_t) noexcept
+              : span{}
+            {}
+            constexpr span(pointer ptr, index_type count) noexcept
+              : size_base_t{(RANGES_ENSURE(count >= 0),
+                    RANGES_ENSURE(N == dynamic_extent || count == N),
+                    count)}
+              , ptr_{(RANGES_ENSURE(count == 0 || ptr != nullptr), ptr)}
+            {}
+            constexpr span(pointer first, pointer last) noexcept
+              : span{first, last - first}
+            {}
+            // EXTENSION: disambiguate initialization from "pointer, small-literal"
+            template<typename Int,
+                CONCEPT_REQUIRES_(SignedIntegral<Int>())>
+            constexpr span(pointer first, Int count) noexcept
+              : span{first, index_type{count}}
+            {}
+
+
+            template<typename Rng>
+            using ConversionConstraint =
+                meta::bool_<SizedRange<Rng>::value && ContiguousRange<Rng>::value &&
+                    detail::is_more_cv_than<T, concepts::ContiguousRange::datum_t<Rng>>::value &&
+                    (N == dynamic_extent ||
+                        range_cardinality<Rng>::value == finite ||
+                        range_cardinality<Rng>::value == N)>;
+
+            template<typename Rng,
+                CONCEPT_REQUIRES_(ConversionConstraint<Rng>())>
+            constexpr span(Rng && rng)
+              : span{ranges::data(rng), static_cast<index_type>(ranges::size(rng))}
+            {}
+
+            // [span.sub], span subviews
+            template<std::ptrdiff_t Count,
+                CONCEPT_REQUIRES_(Count >= 0 && (N == dynamic_extent || Count <= N))>
+            constexpr span<T, Count> first() const noexcept
+            {
+                return RANGES_ENSURE(Count <= size()),
+                    span<T, Count>{data(), Count};
+            }
+            template<std::ptrdiff_t Count,
+                CONCEPT_REQUIRES_(Count >= 0 && (N == dynamic_extent || Count <= N))>
+            constexpr span<T, Count> last() const noexcept
+            {
+                return RANGES_ENSURE(Count <= size()),
+                    span<T, Count>{data() + size() - Count, Count};
+            }
+            template<std::ptrdiff_t Offset, std::ptrdiff_t Count = dynamic_extent,
+                CONCEPT_REQUIRES_(Offset >= 0 && (Count == dynamic_extent ||
+                    (Count >= 0 && (N == dynamic_extent || Offset + Count <= N))))>
+            constexpr span<T, Count> subspan() const noexcept
```

Comment:
Mostly just being a Scrooge. I don't understand why `span` is better than the more general `iterator_range`, and it bugs me the way C++ chooses names that ignore decades of CS convention. Right now I'm feeling like `span` is humbug and nonsense.